include/quic/datagram.hpp

```diff
@@ -9,9 +9,45 @@ namespace oxen::quic
     class Connection;
     class Endpoint;
     class Stream;
+    class connection_interface;
+    struct ConnectionID;
+
+    struct dgram_interface
+    {
+        dgram_interface(Connection& c);
+        connection_interface& ci;
+
+        const ConnectionID& conn_id() const;
+
+        std::shared_ptr<connection_interface> get_conn_interface();
+
+        template <
+                typename CharType,
+                std::enable_if_t<sizeof(CharType) == 1 && !std::is_same_v<CharType, std::byte>, int> = 0>
+        void reply(std::basic_string_view<CharType> data, std::shared_ptr<void> keep_alive = nullptr)
+        {
+            reply(convert_sv<std::byte>(data), std::move(keep_alive));
+        }
+
+        template <typename Char, std::enable_if_t<sizeof(Char) == 1, int> = 0>
+        void send_datagram(std::vector<Char>&& buf)
+        {
+            reply(std::basic_string_view<Char>{buf.data(), buf.size()}, std::make_shared<std::vector<Char>>(std::move(buf)));
+        }
+
+        template <typename CharType>
+        void reply(std::basic_string<CharType>&& data)
+        {
+            auto keep_alive = std::make_shared<std::basic_string<CharType>>(std::move(data));
+            std::basic_string_view<CharType> view{*keep_alive};
+            reply(view, std::move(keep_alive));
+        }
+
+        void reply(bstring_view data, std::shared_ptr<void> keep_alive = nullptr);
```

Comment:
This should probably be renamed as it is used for a non-reply send as well as a reply.

Reply:
That's valid, but from the perspective of the callback, having "reply" somewhere makes it clear that it's directly back along that connection

From that datagram interface, the user has a getter method to retrieve the connection interface and call `send_datagram` from there

Reply:
I think `reply` is fine, but should the `send_datagram` in here taking a vector of char's also be renamed to `reply`?

Reply:
That's a good point, I just used the same templating we had on connection_interface and stream for consistency. Maybe someone wants to send a big reply?